### ISSUE TITLE
Protocol version is not failing back to 3.1

### ISSUE THREAD

**Issue by bkupidura** [eclipse/paho.mqtt.python#82]
Software:
Ubuntu 14.04 on rassberry PI 2
paho-mqtt 1.2
mosquitto 0.15-2ubuntu1
mosquitto-clients 0.15-2ubuntu1
libmosquitto0 0.15-2ubuntu1

Mosquitto on Ubuntu 14.04 supports only MQTT 3.1, failback is handled by paho/mqtt/client.py:_handle_connack but currently this is not executed as _sock.recv(1) from _packet_read returns 0 length string.

if self._in_packet['command'] == 0:
[..]
  if len(command) == 0:
    return 1

After adding a few additional self._easy_log:
2016-06-20 13:30:51,056 - INFO - mqtt - _packet_read
2016-06-20 13:30:51,057 - INFO - mqtt - self._in_packet['command'] == 0
2016-06-20 13:30:51,058 - INFO - mqtt - self._in_packet['command'] try
2016-06-20 13:30:51,058 - INFO - mqtt - self._in_packet['command'] else
2016-06-20 13:30:51,059 - INFO - mqtt - self._in_packet['command'] len(command): 0

Broker logs:
1466422205: New connection from 127.0.0.1.
1466422205: Invalid protocol "MQTT" in CONNECT from 127.0.0.1.
1466422205: Socket read error on client (null), disconnecting.

To reproduce issue you can use publish.single:

import paho.mqtt.publish as publish

publish.single('test/test', payload='test', retain=False, qos=0, hostname='localhost')

Moved from: https://bugs.eclipse.org/bugs/show_bug.cgi?id=496386

**Comment by ralight** [comment]
I'm afraid the only useful way to be able to fail back to mqtt v3.1 is for the broker to return CONNACK with rc=1 (connection refused, protocol version). Mosquitto 0.15 doesn't do this. If the client attempted to reconnect with mqtt v3.1 when there was a read failure that wouldn't be very useful behaviour for most people.

I'm afraid my response here is that you should upgrade to a more recent version of mosquitto. 0.15 is more than 4 years old now.

**Comment by bkupidura** [comment]
As Ubuntu 16.04 is available, this is probably best approach.

Workaround for people with old mosquitto server:

client = paho.Client(protocol=paho.MQTTv31)

**Comment by ralight** [comment]
Yes both of those are reasonable responses, but either way you'll get better support by using the repositories provided by the project.
